kura/org.eclipse.kura.wire.camel/src/main/java/org/eclipse/kura/wire/camel/CamelProduce.java

```diff
@@ -15,18 +15,46 @@
 import static org.apache.camel.builder.DefaultFluentProducerTemplate.on;
 
 import org.apache.camel.CamelContext;
+import org.apache.camel.FluentProducerTemplate;
 import org.eclipse.kura.wire.WireEnvelope;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class CamelProduce extends AbstractReceiverWireComponent {
 
+    private static final Logger logger = LoggerFactory.getLogger(CamelProduce.class);
+
+    private FluentProducerTemplate template = null;
+
     @Override
     protected void processReceive(final CamelContext context, final String endpointUri, final WireEnvelope envelope)
             throws Exception {
-
-        on(context) //
+        if (template == null) {
```

Comment:
I think the logic itself is good. However, I also think there can be a race condition when multiple threads are active.

Assuming that the `if == null` check will most often be false, and setting the template will be much less frequent than getting it, a "read/write lock" makes sense. Just beware, you can't upgrade it from a read lock to a write lock. So you need to unlock the read lock first, then write lock, and check again (if some other thread already created the template).

Reply:
Add readWrite lock for template